rust/gitxet/tests/integration_tests/test_integration_scenarios.sh

```diff
@@ -0,0 +1,516 @@
+cp d3.dat ../d3.dat
+git add d3.dat
+git commit -a -m "Added d3.dat"
+git push origin
+popd
+
+# STATE:
+#  origin_a: XET, D1 + D2
+#    repo_a_1: XET, D1 + D2
+#    repo_a_2: XET, D1 + D2
+#    repo_a_3: XET, D1
+#    repo_a_4: XET, D1 + D2
+#    repo_a_5: NO XET, no data
+#    repo_a_X1: XET, no data
+#    repo_a_X2: XET, no data
+#  origin_b1: NO XET, no data
+#  origin_b2: NO XET, no data
+#  origin_x1: XET, D2 + D1 + D3 
+#    repo_x1_1: XET, D1 + D2 + D3
+#    repo_x1_2: XET, D2
+#  origin_x2: XET, no data 
+
+# Now, simulate a pr merge!   
+simulate_pr_merge origin_x1 origin_a
+
+# See if that works with a clone
+git clone origin_a repo_a_6
+assert_files_equal repo_a_6/d1.dat d1.dat 
+assert_files_equal repo_a_6/d2.dat d2.dat 
+assert_files_equal repo_a_6/d3.dat d3.dat 
+
+# See if that works with a merge 
+pushd repo_a_1
+git fetch origin
+git merge --no-edit origin/main 
+popd
+assert_files_equal repo_a_1/d1.dat d1.dat 
+assert_files_equal repo_a_1/d2.dat d2.dat 
+assert_files_equal repo_a_1/d3.dat d3.dat 
+
+# See if it works with a merge from something older. 
+pushd repo_a_2
+git fetch origin
+git merge --no-edit origin/main 
+popd
+assert_files_equal repo_a_2/d1.dat d1.dat 
+assert_files_equal repo_a_2/d2.dat d2.dat 
+assert_files_equal repo_a_2/d3.dat d3.dat 
+
+# See if it works with a merge from an old non-xet enabled repo.
+pushd repo_a_5
+git fetch origin
+git merge --no-edit origin/main 
+popd
+assert_files_equal repo_a_5/d1.dat d1.dat
+assert_files_equal repo_a_5/d2.dat d2.dat
+assert_files_equal repo_a_5/d3.dat d3.dat
+
+
+# Now, see if repo_x1_2 (XET, D2) can get the new things from origin_a since it's a clone of origin_x1
+pushd repo_x1_2
+git remote add upstream ../origin_a
+git fetch upstream
+git merge --no-edit upstream/main
+popd
+assert_files_equal repo_x1_2/d1.dat d1.dat
+assert_files_equal repo_x1_2/d2.dat d2.dat
+assert_files_equal repo_x1_2/d3.dat d3.dat
+
+# Now see if clones of the other forks can pull from any of the xet enabled remotes and work fine
+git clone origin_b1 repo_b1_1
+pushd repo_b1_1
+git remote add upstream ../origin_a
+git fetch upstream
+git merge --no-edit upstream/main
+git push origin
+popd
+assert_files_equal repo_b1_1/d1.dat d1.dat
+assert_files_equal repo_b1_1/d2.dat d2.dat
+assert_files_equal repo_b1_1/d3.dat d3.dat
+
+# Now, make sure that origin can be cloned and have it work fine. 
+git clone origin_b1 repo_b1_2
+assert_files_equal repo_b1_2/d1.dat d1.dat
+assert_files_equal repo_b1_2/d2.dat d2.dat
+assert_files_equal repo_b1_2/d3.dat d3.dat
+
+# Work with a sideways merge then a PR push across that?
+git clone origin_b2 repo_b2_1
+pushd repo_b2_1
+git remote add upstream ../repo_a_4
+git fetch upstream 
+git merge --no-edit upstream/main
+create_data_file d4.dat 1000
+cp d4.dat ../
+git add d4.dat
+git commit -a -m "Added d4.dat"
+[[ ! -z $(git show HEAD:d4.dat | grep "xet version") ]] || die "d4.dat is not a pointer file."
+
+# Push to origin_b2
+git push origin
+popd
+
```

Comment:
A state chart would be very helpful at this point.

Reply:
Sounds good, will fill it out.  Or try to separate out some things....